### ISSUE TITLE
Replace if display-graphic-p with the proper face display style

### ISSUE THREAD

**Issue by jasonm23** [emacsfodder/emacs-theme-darktooth#1]
Add this issue to gruvbox too

**Comment by jasonm23** [comment]
@purcell - it would be great if something as terse as the `display-graphic-p` was available. The defface syntax is so bloody clunky.

It would be a pretty big benefit if people could define colors in a very simple way like this and apply them.

Any ideas?

**Comment by purcell** [comment]
Yes, I have some thoughts on this. Requires some macro magic. It's definitely a puzzle which needs solving, also for my own solarized themes.

**Comment by purcell** [comment]
Here's a sketch of one way to do it:

``` el
(defconst darktooth-colors
  '((darktooth-dark0 . ((((type graphic)) .  "#1d2021") (t . "color-234")))
    (darktooth-dark1 . ((((type graphic)) .  "#3c3836") (t . "color-237")))))

(defun darktooth-face-spec (spec)
  (mapcar
   (lambda (display)
     (list display (mapcar
                    (lambda (element)
                      (let* ((entry (assoc element darktooth-colors))
                             (variant (assoc display (cdr entry)))
                             (subst (cdr variant)))
                        (when (and entry (null subst))
                          (error "No %s entry for %s" display element))
                        (or subst element)))
                    spec)))
   '(((type graphic)) t)))

(deftheme darktooth)
(custom-theme-set-faces
 'darktooth
 ;; UI
 `(fringe                            ,(darktooth-face-spec '(:background darktooth-dark0)))
 `(hl-line                           ,(darktooth-face-spec '(:background darktooth-dark1)))
)

```

**Comment by jasonm23** [comment]
That's very nice, thanks Steve.

**Comment by jasonm23** [comment]
Added to branch **refit-palette**  I realised that ansi color names vector will pose a slight problem.

Obviously the way I've set it there is wrong, but I think it'll be easier to just make a dark tooth set ansi vector instead.

**Comment by jasonm23** [comment]
https://github.com/emacsfodder/emacs-theme-darktooth/blob/refit-palette/darktooth-theme.el

**Comment by jasonm23** [comment]
darktooth-set-ansi function that is.

**Comment by jasonm23** [comment]
Also, it occurred to me that it would be trivial to express the palette constant in more simplistic terms. E.g.

```
 (name . ("gui color" "xterm"))
```

and then simply transform the entry in the darktooth-face-spec.

**Comment by jasonm23** [comment]
Or just do the whole list for each lookup.  Granted it's more costly but the list will never be very long, and the edit-ability / parse-ability of the palette will be improved.

I'm trying to slot this spec into the vapourware "advanced" mode of my theme editor.  It would be nice if we could come up with a community viable solution.

**Comment by purcell** [comment]
Yes, there would be more concise ways to do this. I might bundle up something like this into a package.

P.S. There's no need to use such a short function name (dt//spec) in order to keep things tidy: you can just do something like this:

``` el
(apply #'custom-theme-set-faces
       'darktooth
       ;; UI
       (mapcar
        (lambda (entry)
          (list (car entry) (dt//spec (cdr entry))))
        '((default (:foreground darktooth-light0 :background darktooth-dark0))
          (cursor (:background darktooth-light0))
          (mode-line (:box nil :foreground darktooth-light1 :background darktooth-dark0_hard))
          (mode-line-inactive (:box nil :background darktooth-dark2 :foreground darktooth-light4))
          (fringe (:background darktooth-dark0))
          (linum (:foreground darktooth-dark4))
          (hl-line (:background darktooth-dark1))
          (region (:background darktooth-dark2))
          (secondary-selection (:background darktooth-dark1))
          (minibuffer-prompt (:background darktooth-dark0 :foreground darktooth-turquoise4 :bold nil)))))
```

**Comment by jasonm23** [comment]
> P.S. There's no need to use such a short function name (dt//spec) in order to keep things tidy: you can just do something like this:

Very true, and very nice.

Bundling these tools into a package would be great, anything that reduces bad copypasta in themes is good.

**Comment by jasonm23** [comment]
Hey @purcell if you get a moment, please have a look at this https://github.com/emacsfodder/emacs-theme-darktooth/blob/refit-palette/darktooth-theme.el

I'm not sure what I'm doing wrong...!

**Comment by sebastiansturm** [comment]
A helper macro that automatically extracts colors from a predefined palette would be doubly useful, because we could use that palette to loop over all faces defined in `(face-list)` and replace their color values by the best approximations the palette has to offer (however "best" may be defined). That way, the theme could perhaps offer a better baseline for all those faces that have not been manually fine-tuned by the theme author? My elisp-fu is not too great, but I have cobbled together a prototype that seems to work, more or less, see https://gist.github.com/sebastiansturm/284d872f931e005fa83a

What do you think (in principle, of course there's still much to do if it's supposed to turn into a useful package)?

**Comment by purcell** [comment]
@jasonm23 Sorry, haven't had chance to review this yet, or work on things further myself.

@sebastiansturm The problem with looping over `face-list` is that it doesn't return faces which haven't been defined yet, which means that the time at which the theme is loaded becomes significant. That wouldn't be ideal.

**Comment by sebastiansturm** [comment]
@purcell I know, but I thought it should be possible to cover that by defining advice for `custom-declare-face`, no?

**Comment by purcell** [comment]
@sebastiansturm Perhaps. Is that what `defface` calls? It sounds a little high on the "magic scale" for my tastes.

**Comment by sebastiansturm** [comment]
@purcell as far as I can tell that is what `defface` calls, yes. I understand your objection, and I'm not even sure I would use the advice'd `defface` myself (personally I'd find it acceptable to occasionally call `load-theme` again after new faces have been defined). Another possible problem is that, if autotheming achieves an acceptable level of visual consistency, it might reduce the incentive for theme users to create pull requests with their hand-tuned face definitions. Maybe an acceptable compromise would be to make the autogeneration of facespecs optional (turned off by default), and add a helper function that can be used to generate equivalent elisp code? That way the effort necessary to extend a theme to one's favorite modes (and to create a suitable pull request) could be much reduced I guess?

**Comment by jasonm23** [comment]
Color distance to say xterm is already done somewhere (although I can't recall where right now, on the bus!)

I'll dig it up.

**Comment by jasonm23** [comment]
@sebastiansturm @purcell my feeling is that something as invasive / magic as advising `defface` isn't a great way to go.  My initial thought was for a requirable library for theme authors to include, so that the following were provided:
1. Theme Palette which is easy to use / modify
2. Easy to configure palette color alternatives (either manually provided by the author or calculated from color distance to xterm)
3.  Straightforward method to provide things such as Ansi colors vector.

I think keeping these tools simple and un-invasive will help them become adopted by theme authors.

**Comment by sebastiansturm** [comment]
@jasonm23 that's fine with me, let's forget about the `defface` stuff. But a helper function that generates a usable elisp template would be unobtrusive, I think, and could encourage more users to contribute to the theme?

**Comment by jasonm23** [comment]
Ideally we can extract and improve on a re-usable library/toolkit for theme authors.

**Comment by sebastiansturm** [comment]
I'd like to revive this issue with a new proposal, please see the following gist: https://gist.github.com/sebastiansturm/9765854a3a65e2b336dc
This provides a macro `autothemer-defautotheme` that accepts a list of colors and a list of stripped-down face specs to create an emacs theme, e.g.

```
(autothemer-defautotheme 
               sometheme "somethemename"
               ((red "#880011") 
                (orange "DarkGoldenrod4")
                (blue "#001199"))
               ((flycheck-error (:underline t :foreground red :weight 'demibold))
                (button (:box nil :background blue))
                ;; etc.
                )
               ;; execute code that uses red, orange, blue
               )
```

would create a theme named "sometheme" that customizes `flycheck-error` and `button`. It does not automatically customize unspecified faces as I had previously proposed, but it provides an interactive function called `autothemer-generate-facespecs` that iterates over all known faces not explicitly specified within `autothemer-defautotheme`, generates approximate customizations using the theme color palette and displays these definition templates in a temporary elisp buffer. In my mind, that still leaves customization to the user, but makes it much more convenient.
As an elisp newbie, my coding style may be atrocious and currently there is no support for terminal colors (although that should be easy to add).
I'd just like to know whether you'd support the concept, or if you're already working on something better (if so, is there something I can download and use already?)

**Comment by jasonm23** [comment]
I'd support the concept.

**Comment by sebastiansturm** [comment]
glad to hear that! I'll set up a github repo over the weekend, then I hope we can beat the thing into shape together, put it on melpa and discuss how best to add terminal support (I have zero experience with that, as I always use the GUI frontend).

**Comment by jasonm23** [comment]
Terminal would be relatively trivial.

We would just run an Emacs process in batch mode.

**Comment by sebastiansturm** [comment]
hm, I'm not sure what you mean. I was referring to the definition of different face customizations for different terminal types. Personally, I only use the Cocoa frontend, so all of my facespecs read something like `(face ((t (... actual specs ...))))`, but some themes (like emacs-solarized, for instance) go to great lengths to degrade gracefully on less capable terminal types (a bit similar to `emacs-theme-darktooth`'s distinction between graphical and non-graphical terminals, but with even more categories inbetween). Which is of course a good thing and should be encouraged, but I'm still uncertain how best to add this facility to autothemer.
I have now pushed a working version (at least it seems to work on my machine) to https://github.com/sebastiansturm/autothemer

**Comment by jasonm23** [comment]
Oh I see 

Well we should use the _proper_ defface style with class colors etc.

**Comment by sebastiansturm** [comment]
I have implemented support for different terminal types in autothemer; seems to work, although I haven't yet tested whether less capable terminals really use the right definitions. I have also created a fork of emacs-theme-darktooth that uses autothemer and that can automatically generate face spec templates using `autothemer-generate-templates`. Would be great if you could check if it works for you, too. If so, I'll try to turn autothemer into a proper melpa package sometime next week.

**Comment by jasonm23** [comment]
Fixed by 0.3.0
